apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs

```diff
@@ -46,6 +116,159 @@ private SpannerConversionOptions(bool useDBNull)
+            }
+
+            // Validate that mapping matches one of the pre-defined values.
+            var invalidMappings = mappings.Except(AllowedSpannerToClrTypeMappings, StringComparer.OrdinalIgnoreCase);
+            if (invalidMappings.Any())
+            {
+                throw new ArgumentException($"The following mappings are invalid:'{string.Join(",", invalidMappings)}' for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            // Check multiples for each type.
+            // Currently, we have Float64 and Date to check.
+            if (mappings.Count(j => j.StartsWith("Float64", StringComparison.OrdinalIgnoreCase)) > 1)
+            {
+                throw new ArgumentException($"'{input}' is not a valid value as multiple mappings from Float64 to CLR type are provided for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            if (mappings.Count(j => j.StartsWith("Date", StringComparison.OrdinalIgnoreCase)) > 1)
+            {
+                throw new ArgumentException($"'{input}' is not a valid value as multiple mappings from Date to CLR type are provided for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            // If we reach here, all is well.
+            foreach (var mapping in mappings)
+            {
+                // Discard pattern, primarily to simplify code using switch expressions.
+                _ = mapping switch
+                {
+                    Float64ToDecimal => options.Float64ToConfiguredClrType = typeof(decimal),
+                    Float64ToDouble => options.Float64ToConfiguredClrType = typeof(double),
+                    Float64ToSingle => options.Float64ToConfiguredClrType = typeof(float),
+                    Float64ToSpannerNumeric => options.Float64ToConfiguredClrType = typeof(SpannerNumeric),
+                    Float64ToPgNumeric => options.Float64ToConfiguredClrType = typeof(PgNumeric),
+                    DateToDateTime => options.DateToConfiguredClrType = typeof(DateTime),
+                    DateToSpannerDate => options.DateToConfiguredClrType = typeof(SpannerDate),
+                    // We shouldn't come here as validations are in place. We don't have a known mapping.
+                    _ => default
+                };
+            }
+        }
+
+        internal void SetClrToSpannerTypeDefaults()
+        {
+            DecimalToConfiguredSpannerType = SpannerDbType.Float64;
+            DateTimeToConfiguredSpannerType = SpannerDbType.Timestamp;
+        }
+
+        internal void SetSpannerToClrTypeDefaults()
+        {
+            DateToConfiguredClrType = typeof(DateTime);
+            Float64ToConfiguredClrType = typeof(double);
+        }
+
+        private void ValidateAndParseClrToSpannerTypeMappings(string input, SpannerConversionOptions options)
```

Comment:
Ideally, I'd like to reduce the duplication between this and the other method - although it's entirely possible that abstracting the infrastructure will be larger than the current code.